### ISSUE TITLE
Split API Call Block into Separate Pre/Post Call Blocks

### ISSUE THREAD

**Issue by dustin-lunarg** [LunarG/gfxreconstruct#190]
The current capture behavior us to record both the input and output/return values for an API call in single packet, which is written to the capture file after the API call has returned.  This behavior could be modified to split the input and output values into separate packets, with inputs written before the API call and outputs written after the API call.

This could be useful for capturing crashes where the crash happened in the API call due to invalid inputs.  With the current behavior of logging the packet after the API call, the packet would not be written to the file in the case of a crash.  Note that this is only true for API calls with return values and/or output parameters.  For API calls that only have input parameters, the packet is logged before the API call.

In theory, this could also be used for basic synchronization with multi-threaded replay.  Packets captured in the following order would indicate that threads 1 and 2 executed API calls at roughly the same time, while thread 3 executed its API call after thread 1 and 2 completed their calls:
* Thread 1 Input
* Thread 2 Input
* Thread 2 Output
* Thread 1 Output
* Thread 3 Input
* Thread 3 Output

The replay tool would interpret this to mean that the calls from threads 1 and 2 could be executed in parallel by worker threads, but thread 3's call should not be executed until the other calls complete.  Multi-threaded replay would then follow a pattern where it reads an Input packet from the file and dispatches that packet to a worker thread.  It would continue reading and dispatching Input packets until it encounters an Output packet.  When encountering an Output packet, the dispatch thread would block until the associated worker thread completes its API call and commits any resulting state changes (eg. mappings for newly created handles) before continuing to read and process packets from the capture file.

Some downsides to this approach are that every API call, even calls without return values or output parameters, would need an Output packet, and that it complicates single threaded replay.  For single threaded replay, Input packets would need to be queued until the associated Output packet is read.  To maintain the proper API call order, queue entries can only be processed when the item at the front of the queue is complete.

### Edit 2023-03-30 by @andrew-lunarg 
Looks like Dustin did some work on this: https://github.com/LunarG/gfxreconstruct/compare/dev...dustin-lunarg:gfxreconstruct:dustin_separate_in_out_packets

**Comment by per-mathisen-arm** [comment]
For true multi-threaded replay it would be better to read from multiple input blocks/files, one per thread, to avoid this problem entirely. When thinking about redoing packets, it would be good to split compression packets and data packets apart, as what is a good size for a compression packet is probably not a single data packet, and merging these two into a single packet type seems both conceptually wrong and inefficient.

**Comment by andrew-lunarg** [comment]
Dustin seems to have been using the terms "packet" and "block" interchangeably (as does our code a bit), whereas you seem to be calling a per-thread trace a "block" Per. Your points are interesting but seem orthogonal to Dustin's reasons for making this issue.
Can we agree to use the term block in its dominant GFXR meaning \* in this topic?

I think your points are:
1. Capturing separate trace files, one per thread has performance advantages.
1. Decoupling units of compressed data from the blocks allows higher performance and better compression ratios (I know your own vk capture tool `lavatube` has a compression thread and a file write thread per captured thread so I imagine you have tuned the compression buffer size to keep that pipeline flowing).

Dustin seemed concerned about:
1. Capturing the inputs to an API call before the call is taken so they are available in case the call crashes.
1. Preserving the order that calls were made into the API at capture time _across all threads_ when a trace is replayed on multiple threads.

On the second point from Dustin, one reason for this is in the case that the application used its own thread synchronisation rather than purely Vulkan mechanisms, this order may matter for correctness. Examples might be thread one submits a bunch of work, then notifies waiting thread two when it knows that work is complete. Thread two wakes and deletes a bunch of Vulkan handles to resources that it now knows is not in flight on the GPU. If we had just blindly replayed the traces for each thread as fast as they will go, thread two can delete resources still in flight on the GPU.

I think your points are very good ones but perhaps not related to this issue directly. Please do let us know here if you've got solutions from your prototype to the problems Dustin was laying out above. We should continue the conversation about your points in the previous message elsewhere if I have understood them correctly or please let me know if I missed something you were trying to get across.

\*
```
// framework/format/format.h
enum BlockType : uint32_t
{
    kUnknownBlock                = 0,
    kFrameMarkerBlock            = 1, // Marker to denote frame status, such as the start or end of a frame.
    kStateMarkerBlock            = 2, // Marker to denote state snapshot status, such as the start or end of a state snapshot.
    kMetaDataBlock               = 3,
    kFunctionCallBlock           = 4,
    kAnnotation                  = 5,
    kMethodCallBlock             = 6,
    kCompressedMetaDataBlock     = MakeCompressedBlockType(kMetaDataBlock),
    kCompressedFunctionCallBlock = MakeCompressedBlockType(kFunctionCallBlock),
    kCompressedMethodCallBlock   = MakeCompressedBlockType(kMethodCallBlock),
};
```

**Comment by per-mathisen-arm** [comment]
I don't think this is purely a matter of performance, though. Effectively replaying multi-threaded content also means being able to reproduce the original workload where things actually ran in parallel, which in turn requires a certain amount of performance. One lesson learned from implementing this in patrace was that if you don't get the basic synchronization design right, your multi-threaded replay can end up slower than single-threaded and for that reason be less parallelized when the work ends up on the GPU. But yes, should probably continue this discussion elsewhere :-)

My point is that I don't see how the above is actually solving multi-threading. I haven't thought about the debugging app crashes use case much (it is not what we do), but couldn't you instead use the signal handler here and write out all the relevant information through a callstack to the log instead?

**Comment by andrew-lunarg** [comment]
That all makes a lot of sense, thanks :).

Given your PPA focus I'm guessing that being able to light up as much as possible of a pre-silicon (model/simulation/emulation/pick your group's favourite term) would guide a lot of what you have done. Other use cases might care more about optimising applications on released products and have a full system profiler enabled during capture and replay. There could be tradeoffs there from adding a compute thread for compression and IO thread for every application thread calling through the captured API. 

But yes, it doesn't make sense to keep scattering this discussion over old issues.
